
With the separation of pagefault_{disable,enable}() from the preempt_count
a previously overlooked dependancy became painfully clear.

kmap_atomic() is per cpu and relies not only on disabling the pagefault 
handler, but really needs preemption disabled too.

make this explicit now - so that we can change pagefault_disable().

Signed-off-by: Peter Zijlstra <a.p.zijlstra@chello.nl>
---
 arch/i386/mm/highmem.c    |    4 +++-
 arch/mips/mm/highmem.c    |    5 ++++-
 arch/sparc/mm/highmem.c   |    4 +++-
 include/asm-frv/highmem.h |    2 ++
 include/asm-ppc/highmem.h |    4 +++-
 5 files changed, 15 insertions(+), 4 deletions(-)

Index: linux-2.6.23.1-rt5/arch/i386/mm/highmem.c
===================================================================
--- linux-2.6.23.1-rt5.orig/arch/i386/mm/highmem.c
+++ linux-2.6.23.1-rt5/arch/i386/mm/highmem.c
@@ -51,7 +51,7 @@ void *__kmap_atomic_prot(struct page *pa
 	enum fixed_addresses idx;
 	unsigned long vaddr;
 
-	/* even !CONFIG_PREEMPT needs this, for in_atomic in do_page_fault */
+	preempt_disable();
 	pagefault_disable();
 
 	if (!PageHighMem(page))
@@ -93,6 +93,7 @@ void __kunmap_atomic(void *kvaddr, enum 
 
 	arch_flush_lazy_mmu_mode();
 	pagefault_enable();
+	preempt_enable();
 }
 
 /* This is the same as kmap_atomic() but can map memory that doesn't
@@ -103,6 +104,7 @@ void *__kmap_atomic_pfn(unsigned long pf
 	enum fixed_addresses idx;
 	unsigned long vaddr;
 
+	preempt_disable();
 	pagefault_disable();
 
 	idx = type + KM_TYPE_NR*smp_processor_id();
Index: linux-2.6.23.1-rt5/arch/mips/mm/highmem.c
===================================================================
--- linux-2.6.23.1-rt5.orig/arch/mips/mm/highmem.c
+++ linux-2.6.23.1-rt5/arch/mips/mm/highmem.c
@@ -38,7 +38,7 @@ void *__kmap_atomic(struct page *page, e
 	enum fixed_addresses idx;
 	unsigned long vaddr;
 
-	/* even !CONFIG_PREEMPT needs this, for in_atomic in do_page_fault */
+	preempt_disable();
 	pagefault_disable();
 	if (!PageHighMem(page))
 		return page_address(page);
@@ -63,6 +63,7 @@ void __kunmap_atomic(void *kvaddr, enum 
 
 	if (vaddr < FIXADDR_START) { // FIXME
 		pagefault_enable();
+		preempt_enable();
 		return;
 	}
 
@@ -78,6 +79,7 @@ void __kunmap_atomic(void *kvaddr, enum 
 #endif
 
 	pagefault_enable();
+	preempt_enable();
 }
 
 /*
@@ -89,6 +91,7 @@ void *kmap_atomic_pfn(unsigned long pfn,
 	enum fixed_addresses idx;
 	unsigned long vaddr;
 
+	preempt_disable();
 	pagefault_disable();
 
 	idx = type + KM_TYPE_NR*smp_processor_id();
Index: linux-2.6.23.1-rt5/arch/sparc/mm/highmem.c
===================================================================
--- linux-2.6.23.1-rt5.orig/arch/sparc/mm/highmem.c
+++ linux-2.6.23.1-rt5/arch/sparc/mm/highmem.c
@@ -34,7 +34,7 @@ void *kmap_atomic(struct page *page, enu
 	unsigned long idx;
 	unsigned long vaddr;
 
-	/* even !CONFIG_PREEMPT needs this, for in_atomic in do_page_fault */
+	preempt_disable();
 	pagefault_disable();
 	if (!PageHighMem(page))
 		return page_address(page);
@@ -71,6 +71,7 @@ void kunmap_atomic(void *kvaddr, enum km
 
 	if (vaddr < FIXADDR_START) { // FIXME
 		pagefault_enable();
+		preempt_enable();
 		return;
 	}
 
@@ -97,6 +98,7 @@ void kunmap_atomic(void *kvaddr, enum km
 #endif
 
 	pagefault_enable();
+	preempt_enable();
 }
 
 /* We may be fed a pagetable here by ptep_to_xxx and others. */
Index: linux-2.6.23.1-rt5/include/asm-frv/highmem.h
===================================================================
--- linux-2.6.23.1-rt5.orig/include/asm-frv/highmem.h
+++ linux-2.6.23.1-rt5/include/asm-frv/highmem.h
@@ -115,6 +115,7 @@ static inline void *kmap_atomic(struct p
 {
 	unsigned long paddr;
 
+	preempt_disable();
 	pagefault_disable();
 	paddr = page_to_phys(page);
 
@@ -171,6 +172,7 @@ static inline void kunmap_atomic(void *k
 		BUG();
 	}
 	pagefault_enable();
+	preempt_enable();
 }
 
 #endif /* !__ASSEMBLY__ */
Index: linux-2.6.23.1-rt5/include/asm-ppc/highmem.h
===================================================================
--- linux-2.6.23.1-rt5.orig/include/asm-ppc/highmem.h
+++ linux-2.6.23.1-rt5/include/asm-ppc/highmem.h
@@ -78,7 +78,7 @@ static inline void *kmap_atomic(struct p
 	unsigned int idx;
 	unsigned long vaddr;
 
-	/* even !CONFIG_PREEMPT needs this, for in_atomic in do_page_fault */
+	preempt_disable();
 	pagefault_disable();
 	if (!PageHighMem(page))
 		return page_address(page);
@@ -102,6 +102,7 @@ static inline void kunmap_atomic(void *k
 
 	if (vaddr < KMAP_FIX_BEGIN) { // FIXME
 		pagefault_enable();
+		preempt_enable();
 		return;
 	}
 
@@ -115,6 +116,7 @@ static inline void kunmap_atomic(void *k
 	flush_tlb_page(NULL, vaddr);
 #endif
 	pagefault_enable();
+	preempt_enable();
 }
 
 static inline struct page *kmap_atomic_to_page(void *ptr)
